File: leetcode/typescript/000169.majority-element.ts
function majorityElement(nums: number[]): number {
  let max: [k: number, v: number] = [0, 0]; // tuple: [k, v]
  const map = new Map<number, number>();

  for (let i = 0; i < nums.length; i++) {
    map.set(nums[i], (map.get(nums[i]) || 0) + 1);
  }

  for (const [k, v] of map) {
    if (v > max[1]) {
      max = [k, v];
    }
  }

  return max[0];
}
// Time complexity: O(n)
// Space complexity: O(n)

// Boyer-Moore algorithm
function majorityElement2(nums: number[]): number {
  let curCandidate = nums[0];
  let count = 1;

  for (let i = 1; i < nums.length; i++) {
    if (nums[i] === curCandidate) {
      count++;
    } else {
      count--;
    }

    if (count === 0) {
      i++;
      curCandidate = nums[i];
      count = 1;
    }
  }

  return curCandidate;
}
// Time complexity: O(n)
// Space complexity: O(1)

File: leetcode/python/000169.majority-element.py
from collections import Counter


class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def majorityElement(self, nums: list[int]) -> int:
        count = Counter(nums)
        most_common = max(count.values())
        for k, v in count.items():
            if v == most_common:
                return k
        return 0  # won't happen

    # Time complexity: O(n)
    # Space complexity: O(1)
    def majorityElement1(self, nums: list[int]) -> int:
        count, res = 1, nums[0]
        for n in nums[1:]:
            if count == 0:
                res = n
            if n == res:
                count += 1
            else:
                count -= 1
        return res
